### PR TITLE
[Feature] Nested params scrubber

### PR DESCRIPTION
**Pull request by pawelkomarnicki**
Does scrub deeply-nested hashes :-)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -48,7 +48,6 @@ jobs:
             TEST_FILES="$(circleci tests glob "spec/**/*_spec.rb" | circleci tests split --split-by=timings)"
 
             bundle exec rspec --format progress \
-                            --format RspecJunitFormatter \
```
Comment (tprzedmojski) [inline .circleci/config.yml]: why did you change this?

File: lib/blinkist-airbrake-scrubber/airbrake.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Prepend the original Airbrake module
+module Airbrake
+  class << self
+    prepend Blinkist::Airbrake::Scrubber
+  end
+end
```
Comment (tprzedmojski) [inline lib/blinkist-airbrake-scrubber/airbrake.rb]: I don't like that this lives in a separate file now. There are two problems with that:
1. Is is now hidden, i.e. not in a place where you would immediately look (the main file of the gem)
2. It is load-order dependant. I.e. if you load this file before `Scrubber`, everything breaks.

File: lib/blinkist-airbrake-scrubber/hash.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Add Hash#traverse(&block)
+class Hash
```
Comment (tprzedmojski) [inline lib/blinkist-airbrake-scrubber/hash.rb]: Never ever extend the core classes unless you have a very good reason for it. There is no such reason here!

File: lib/blinkist-airbrake-scrubber/hash.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      elsif value.respond_to?(:to_hash)
+        value = value.to_hash.traverse(&block)
+      end
+
+      # Process key-value pair
+      new_key, new_value = block.call key, value
+      memo[new_key]      = new_value
+
+      memo
+    end
```
Comment (tprzedmojski) [inline lib/blinkist-airbrake-scrubber/hash.rb]: This method is nice, but also insufficient for our purposes, since it fails to catch for example:

```ruby
{ params: [ { email: "a@a.com" }, { email: "b@b.com" } ] }
```

Comment (tprzedmojski) [reply lib/blinkist-airbrake-scrubber/hash.rb]: An alternative (probably a bit buggy):
```ruby
class DeepTraversal
  def initialize(source)
    @source = source
  end

  def traverse(&block)
    recursive_traverse(@source, &block)
  end

  private

  def recursive_traverse(input, &block)
    case input
    when Array
      input.map { |i| recursive_traverse(i, &block) }
    when Hash
      Hash[input.map { |key, value|
        case value
        when String
          [key, block.call(key, value)]
        else
          [key, recursive_traverse(value, &block)]
        end
      }]
    else
      input
    end
  end
end

source = { params: [ { email: "a@a.com" }, { email: "b@b.com" }, {x: "blah"} ], email: "c@c.com", a: 1 }

target = DeepTraversal.new(source).traverse do |key, value|
  if key == :email
    "[Filtered]"
  else
    value
  end
end

pp source
pp target
```

File: .circleci/config.yml (hunk 1 above)
Comment (pawelkomarnicki) [reply .circleci/config.yml]: It was crashing my CircleCI tests (and I don't even know if need JunitFormatter

File: lib/blinkist-airbrake-scrubber/airbrake.rb (hunk 2 above)
Comment (pawelkomarnicki) [reply lib/blinkist-airbrake-scrubber/airbrake.rb]: Good point, I will revert this

File: lib/blinkist-airbrake-scrubber/hash.rb (hunk 4 above)
Comment (pawelkomarnicki) [reply lib/blinkist-airbrake-scrubber/hash.rb]: Okay, but do we anywhere in Airbrake have such case?

Comment (tprzedmojski) [reply lib/blinkist-airbrake-scrubber/hash.rb]: I would bet about 3.50€ that we have something like this hiding somewhere

File: lib/blinkist-airbrake-scrubber/hash.rb (hunk 3 above)
Comment (pawelkomarnicki) [reply lib/blinkist-airbrake-scrubber/hash.rb]: Apart from the possibility to of method name collision, is there really a reason for not doing that? It's like the "poster feature" of Ruby: _you can easily extend any classes_ ;-) (also, I acknowledge your feedback about it and gonna try to find a better solution, too)

File: lib/blinkist-airbrake-scrubber/hash.rb (hunk 4 above)
Comment (pawelkomarnicki) [reply lib/blinkist-airbrake-scrubber/hash.rb]: That's a helluva lot of money, mister! :P But I like this idea, since it would definitely make the scrubbing more sturdy, gonna try to get that :-)

File: lib/blinkist-airbrake-scrubber/hash.rb (hunk 3 above)
Comment (tprzedmojski) [reply lib/blinkist-airbrake-scrubber/hash.rb]: The possibility of name collision was big enough of a problem that ruby language developers tried to solve this with an abomination of a feature called refinements: https://docs.ruby-lang.org/en/2.4.0/syntax/refinements_rdoc.html

> It's like the "poster feature" of Ruby: you can easily extend any classes ;-) 

yes, and it is a bad one from software design perspective. In practice it makes refactoring and debugging much harder.

don't get me wrong: it can be used productively and can be super useful, but you need to have a very good reason for doing so. "Because I can" does not cut it!

Comment (pawelkomarnicki) [reply lib/blinkist-airbrake-scrubber/hash.rb]: Sure, will create a separate class like you suggested with `DeepTraversal`. I wanted a 'quick little fix' with that Hash#traverse method, and I'm not attached to this particular implementation :-)

File: lib/blinkist-airbrake-scrubber/deep_traversal.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# DeepTraversal provides traverse possibility of Hashes
+# Can handle Hash objects with deep nesting, or other nested objects that can be dug deeper (e.g. Array)
+class DeepTraversal
```
Comment (tprzedmojski) [inline lib/blinkist-airbrake-scrubber/deep_traversal.rb]: This needs to be inside `Blinkist::Airbrake::Scrubber` potentially inside `Blinkist::Airbrake::Scrubber::Utils`

File: lib/blinkist-airbrake-scrubber/deep_traversal.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  private
+
+  def recursive_traverse(input, &block)
+    if input.is_a? Array
+      input.map { |i| recursive_traverse(i, &block) }
+
+    elsif input.respond_to?(:to_hash)
+      Hash[input.map { |key, value|
```
Comment (tprzedmojski) [inline lib/blinkist-airbrake-scrubber/deep_traversal.rb]: This is wrong: not everything that responds to `to_hash` will be a hash and will implement `.map` method. A little bit less wrong, but still pretty wrong would be `input.to_hash.map`
